### PR TITLE
PMM-8214 Kubernetes monitoring.

### PR DESCRIPTION
**Pull request by BupycHuk**
https://github.com/Percona-Lab/pmm-submodules/pull/1902
https://github.com/percona/pmm-managed/pull/799
https://github.com/percona-platform/dbaas-controller/pull/201
https://github.com/percona-platform/dbaas-api/pull/149

### CODE REVIEW

**Comment by askomorokhov** [comment]
@BupycHuk, could you please check `packages.dot` - CI fails.

File: utils/logger/zap_logger.go [hunk 1]
```diff
@@ -60,6 +60,11 @@ func (z *ZapLogger) WithField(key string, value interface{}) Logger {
 	return &ZapLogger{l: z.l.With(key, value)}
 }
 
+// SetLevel sets log level.
+func (z *ZapLogger) SetLevel(level Level) {
+	// TODO: implement it.
```
Comment (jprukner) [inline utils/logger/zap_logger.go:65]: Are we gonna ever implement this, or is it just stub so we comply with Logger interface?

File: utils/app/app.go [hunk 2]
```diff
@@ -58,6 +60,7 @@ func Setup(opts *SetupOpts) (*Flags, error) {
 	var flags Flags
 	kingpin.Flag("grpc.addr", "gRPC listen address").Default(":20201").StringVar(&flags.GRPCAddr)
 	kingpin.Flag("debug.addr", "Debug listen address").Default(":20203").StringVar(&flags.DebugAddr)
+	kingpin.Flag("debug", "Enable debug").Envar("PMM_DEBUG").BoolVar(&flags.LogDebug)
```
Comment (jprukner) [inline utils/app/app.go:63]: :+1:

File: service/k8sclient/common/meta.go [hunk 3]
```diff
@@ -68,3 +68,52 @@ type ObjectMeta struct {
...
+// Note:
+// There are two different styles of label selectors used in versioned types:
+// an older style which is represented as just a string in versioned types, and a
+// newer style that is structured.  LabelSelector is an internal representation for the
+// latter style.
+
+// LabelSelector is a label query over a set of resources. The result of matchLabels and
+// matchExpressions are ANDed. An empty label selector matches all objects. A null
+// label selector matches no objects.
+// +structType=atomic
```
[the remark this answers is not in the document]
Comment (jprukner) [reply service/k8sclient/common/meta.go]: is it copy pasted from the operator code? If so, it's fine, I guess. If not, I don't get why such comments are here.

File: deploy/victoriametrics/crs/vmagent_rbac.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  - apiGroups: [""]
+    resources:
+      - namespaces
+      - configmaps
+    verbs: ["get"]
+  - nonResourceURLs: ["/metrics","/metrics/resources"]
+    verbs: ["get"]
+  - apiGroups:
+      - route.openshift.io
+      - image.openshift.io
```
Comment (jprukner) [inline deploy/victoriametrics/crs/vmagent_rbac.yaml:32]: why is it openshift.io?

File: service/k8sclient/common/meta.go [hunk 5]
```diff
@@ -68,3 +68,52 @@ type ObjectMeta struct {
...
+type LabelSelector struct {
+	// matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
+	// map is equivalent to an element of matchExpressions, whose key field is "key", the
+	// operator is "In", and the values array contains only "value". The requirements are ANDed.
+	// +optional
+	MatchLabels map[string]string `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
```
Comment (jprukner) [inline service/k8sclient/common/meta.go:86]: the same for protobuf tag

**Review by jprukner — COMMENTED** [review]
I've got some questions. See them bellow.

File: utils/logger/zap_logger.go (hunk 1 above)
Comment (BupycHuk) [reply utils/logger/zap_logger.go:65]: We will implement it once we start using dbaas-controller in platform.
For now it's just to comply with Logger interface

File: deploy/victoriametrics/crs/vmagent_rbac.yaml (hunk 4 above)
Comment (BupycHuk) [reply deploy/victoriametrics/crs/vmagent_rbac.yaml:32]: copy pasted from victoriametrics operator documentation.

File: service/k8sclient/common/meta.go (hunk 5 above)
Comment (BupycHuk) [reply service/k8sclient/common/meta.go:86]: Good point, will remove

File: service/k8sclient/k8sclient.go [hunk 6]
```diff
@@ -1704,3 +1706,92 @@ func (c *K8sClient) InstallPSMDBOperator(ctx context.Context) error {
...
+		},
+		Spec: monitoring.VMAgentSpec{
+			ServiceScrapeNamespaceSelector: new(common.LabelSelector),
+			ServiceScrapeSelector:          new(common.LabelSelector),
+			PodScrapeNamespaceSelector:     new(common.LabelSelector),
+			PodScrapeSelector:              new(common.LabelSelector),
+			ProbeSelector:                  new(common.LabelSelector),
+			ProbeNamespaceSelector:         new(common.LabelSelector),
+			StaticScrapeSelector:           new(common.LabelSelector),
+			StaticScrapeNamespaceSelector:  new(common.LabelSelector),
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply service/k8sclient/k8sclient.go:1761]: https://github.com/VictoriaMetrics/operator/blob/master/docs/quick-start.MD#VMAgent.
VMAgent has difference behaviour for empty `{}` and nil.

File: utils/logger/zap_logger.go (hunk 1 above)
Comment (jprukner) [reply utils/logger/zap_logger.go:65]: ok

File: service/k8sclient/k8sclient.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jprukner) [reply service/k8sclient/k8sclient.go:1761]: aha, thanks

**Review by jprukner — APPROVED** [review]

**Review by jprukner — APPROVED** [review]

File: service/k8sclient/common/types.go [hunk 7]
```diff
@@ -168,3 +168,22 @@ type NodeSummaryNode struct {
 type NodeSummary struct {
 	Node NodeSummaryNode `json:"node,omitempty"`
 }
+
+// LocalObjectReference contains enough information to let you locate the
+// referenced object inside the same namespace.
+type LocalObjectReference struct {
+	// Name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
```
Comment (askomorokhov) [inline service/k8sclient/common/types.go:177]: Have we started use protobuf? or it is rudimentary here?

File: service/k8sclient/internal/kubectl/kubectl.go [hunk 8]
```diff
@@ -280,6 +280,7 @@ func run(ctx context.Context, kubectlCmd []string, args []string, stdin interfac
 	err := cmd.Run()
 	if err != nil {
 		if strings.Contains(errBuf.String(), "NotFound") {
+			l.Warn(errBuf.String())
```
Comment (askomorokhov) [inline service/k8sclient/internal/kubectl/kubectl.go:283]: errBuf.String() uses several times - looks like it has earned an honor to be a separate variable :)

**Review by askomorokhov — APPROVED** [review]
Looks good to me! 👍

**Pull request merged**